ring/spread_minimizing_token_generator_test.go

```diff
@@ -0,0 +1,558 @@
+		require.Equal(t, allTokens[missingIndex], tokens[i])
+	}
+}
+
+func createTokensForAllInstancesAndZones(t *testing.T, maxInstanceID, tokensPerInstance int) (map[uint32]*instanceInfo, map[string][]uint32) {
+	instanceByToken := make(map[uint32]*instanceInfo, (maxInstanceID+1)*tokensPerInstance*len(zones))
+	tokenSetsByZone := make(map[string][][]uint32, len(zones))
+	for _, zone := range zones {
+		finalInstance := fmt.Sprintf("instance-%s-%d", zone, maxInstanceID)
+		cfg := &SpreadMinimizingConfig{finalInstance, zone}
+		tokenGenerator := createSpreadMinimizingTokenGenerator(t, cfg, zones)
+		tokensByInstance := tokenGenerator.generateTokensByInstanceID()
+		for id, tokens := range tokensByInstance {
+			instInfo := &instanceInfo{
+				InstanceID: fmt.Sprintf("instance-%s-%d", zone, id),
+				Zone:       zone,
+			}
+			for _, token := range tokens {
+				if inst, ok := instanceByToken[token]; ok {
+					err := fmt.Errorf("token %d was already assigned to instance %s", token, inst.InstanceID)
+					panic(err)
+				}
+				instanceByToken[token] = instInfo
+			}
+			allTokens, ok := tokenSetsByZone[zone]
+			if !ok {
+				allTokens = make([][]uint32, 0, maxInstanceID+1)
+			}
+			allTokens = append(allTokens, tokens)
+			tokenSetsByZone[zone] = allTokens
+		}
+	}
+
+	tokensByZone := MergeTokensByZone(tokenSetsByZone)
+	return instanceByToken, tokensByZone
+}
+
+func createSpreadMinimizingTokenGenerator(t testing.TB, cfg *SpreadMinimizingConfig, zones []string) *SpreadMinimizingTokenGenerator {
+	if cfg == nil {
+		cfg = &SpreadMinimizingConfig{"instance-zone-a-10", "zone-a"}
+	}
+	tokenGenerator, err := NewSpreadMinimizingTokenGenerator(*cfg, zones, log.NewLogfmtLogger(os.Stdout))
+	require.NoError(t, err)
+	require.NotNil(t, tokenGenerator)
+	return tokenGenerator
+}
+
+// registeredOwnershipByZone calculates ownership maps grouped by instance id and by zone
+func registeredOwnershipByZone(instancesPerZone int, instanceByToken map[uint32]*instanceInfo, tokensByZone map[string][]uint32) map[string]map[string]float64 {
+	ownershipByInstanceByZone := make(map[string]map[string]float64, len(zones))
+	for zone, tokens := range tokensByZone {
+		ownershipByInstanceByZone[zone] = make(map[string]float64, instancesPerZone)
+		if len(tokens) == 0 {
+			continue
+		}
+		prev := len(tokens) - 1
+		for tk, token := range tokens {
+			ownership := float64(tokenDistance(tokens[prev], token))
+			ownershipByInstanceByZone[zone][instanceByToken[token].InstanceID] += ownership
+			prev = tk
+		}
+	}
+	return ownershipByInstanceByZone
+}
+
+// generateTokensByInstanceID is a modified implementation of SpreadMinimizingTokenGenerator.generateAllTokens that
+// is used in the tests to speed up the generation of tokens, since the generation of tokens for instance with id
+// instanceID already generates the tokens of all other instances with lower ids.
+func (t *SpreadMinimizingTokenGenerator) generateTokensByInstanceID() map[int]Tokens {
```

Comment:
Unfortunately I don't think this is a good idea. In the tests using this, we're not testing the real implementation but a mocked version, so a change in the real implementation which introduces a bug may not being noticed. I strongly recommend to get rid of this.

What can we do instead? We could slightly modify `SpreadMinimizingTokenGenerator.generateAllTokens()` to also return the tokens for all instances (since they were computed). Then this value is ignored by `SpreadMinimizingTokenGenerator.GenerateTokens()`, but in the tests you call `generateAllTokens()` instead of `GenerateTokens()`.